nx/guides/advanced/complex_fft.livemd

```diff
@@ -0,0 +1,297 @@
+
+It will be the sum of two sinusoidal signals, one at 5Hz, and one at 20Hz with the corresponding amplitudes (1, 0.5).
+
+$f(t) = \sin(2\pi*5*t) + \frac12 \sin(2\pi*20*t)$
+
+Suppose we can sample at `fs=50Hz` (meaning 50 samples per second) and our aquisition time is `duration = 1s`.
+
+We build a time series of `fs` points equally spaced with the given `duration` interval with `Nx.linspace`.
+
+For each value of this serie (the discrete time $t$), we will calculate the "hidden" signal $f(t)$ in the module below. This will give us the sampling.
+
+```elixir
+defmodule Signal do
+  import Nx.Defn
+  import Nx.Constants, only: [pi: 0]
+
+  defn source(t) do
+    f1 = 5; f2 = 20;
+    Nx.sin(2 * pi() * f1 * t ) + 1/2 * Nx.sin(2 * pi() * f2 * t)
+  end
+
+  defn sample(opts) do
+    fs = opts[:fs]
+    duration = opts[:duration]
+    sampling = Nx.linspace(0, duration, n: fs, endpoint: false, type: {:f, 32})
+    source(sampling)
+  end
+end
+```
+
+We sample our signal at fs=50Hz during 1s:
+
+```elixir
+fs = 50; duration= 1
+
+sample = Signal.sample(fs: fs, duration: 1)
+```
+
+### Analyse the signal with DFT
+
+The DFT algorithm will **approximates the original signal**. It gives us the following information:
+
+> The number at the index $i$ of the DFT results gives an approximation of the amplitude and phase of the sampled signal at the frequency $i$.
+
+In other words, doing `Nx.fft(sample)` returns a list of numbers indexed by the frequency.
+
+These are complex numbers as they carry the amplitude and phase information.
+
+```elixir
+dft = Nx.fft(sample)
+```
+
+Our frequency resolution - the accuracy of the DFT algorithm - is limited by the Nyquist rate, which is half the sampling rate, thus 25 Hz.
+
+Check [Nyquist](https://en.wikipedia.org/wiki/Nyquist_rate) to read more.
+
+We will therefor limit our study points to the indices below this threshold value by slicing the tensor, thus retain the 25 first ones.
+
+We are interested in the amplitudes only here. This is calculated by taking the norm of each number (`Nx.abs`).
+
+```elixir
+n = Nx.size(dft)
+
+# Nyquist threshold:
+max_freq_index = div(n, 2)
+
+# thus we keep only the meaningfull values and calculate the amplitudes
+amplitudes =  Nx.abs(dft)[0..max_freq_index]
+
+# the frequency bins, "n" of them spaced by fs/n=1 unit:
+frequencies = NxSignal.fft_frequencies(fs, fft_length: n)[0..max_freq_index]
+
+data1 = %{
+  frequencies: Nx.to_list(frequencies),
+  amplitudes: Nx.to_list(amplitudes)
+}
+
+VegaLite.new(width: 700, height: 300)
+|> VegaLite.data_from_values(data1)
+|> VegaLite.mark(:bar)
+|> VegaLite.encode_field(:x, "frequencies",
+  type: :quantitative,
+  title: "frequency (Hz)",
+  scale: [domain: [0, 50]]
+)
+|> VegaLite.encode_field(:y, "amplitudes",
+  type: :quantitative,
+  title: "amplitutde",
+  scale: [domain: [0, 30]]
+)
+```
+
+We see the peaks at 5Hz, 20Hz with their amplitudes (the second is half the first).
+
+This is indeed our incomming signal 🎉
```

Comment:
```suggestion
This is indeed our synthesized signal 🎉
```

Reply:
You mean "synthesized" by  the "original" signal?

Reply:
Yes. As we're synthesizing it instead of sampling from a continuous-time signal. Incoming didn't really resonate with me, and I think original is not really clear.

I'm open to suggestions, and I'm ok with using "original" as well, if you think synthesized is too technical